### PR TITLE
[BEAM-9577] Artifact v2 support for uber jars.

### PR DESCRIPTION
**Pull request by robertwb**
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/ClassLoaderFileSystem.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+public class ClassLoaderFileSystem extends FileSystem<ClassLoaderFileSystem.ClassLoaderResourceId> {
+
+  public static final String SCHEMA = "classpath";
+  private static final String PREFIX = SCHEMA + "://";
+
+  ClassLoaderFileSystem() {}
+
+  @Override
+  protected List<MatchResult> match(List<String> specs) throws IOException {
+    throw new UnsupportedOperationException("Un-globable filesystem.");
```
Comment (ibzib) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/ClassLoaderFileSystem.java:49]:
```suggestion
    throw new UnsupportedOperationException("Un-globbable filesystem.");
```

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/ClassLoaderFileSystem.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.io.fs.CreateOptions;
+import org.apache.beam.sdk.io.fs.MatchResult;
+import org.apache.beam.sdk.io.fs.ResolveOptions;
+import org.apache.beam.sdk.io.fs.ResourceId;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableList;
+
+/** A read-only {@link FileSystem} implementation looking up resources using a ClassLoader. */
```
Comment (ibzib) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/ClassLoaderFileSystem.java:39]: Nice! But I don't see this actually used anywhere?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/ClassLoaderFileSystemTest.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    ReadableByteChannel channel = FileSystems.open(FileSystems.matchNewResource(SOME_CLASS, false));
+    checkIsClass(channel);
+  }
+
+  public void checkIsClass(ReadableByteChannel channel) throws IOException {
+    FileSystems.setDefaultPipelineOptions(PipelineOptionsFactory.create());
+    InputStream inputStream = newInputStream(channel);
+    byte[] magic = new byte[4];
+    inputStream.read(magic);
+    assertArrayEquals(magic, new byte[] {(byte) 0xCA, (byte) 0xFE, (byte) 0xBA, (byte) 0xBE});
```
Comment (ibzib) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/ClassLoaderFileSystemTest.java:55]: How did I never know this? 😆

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/ClassLoaderFileSystem.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+    private final String path;
+
+    private ClassLoaderResourceId(String path) {
+      checkArgument(path.startsWith(PREFIX), path);
+      this.path = path;
+    }
+
+    @Override
+    public ResourceId resolve(String other, ResolveOptions resolveOptions) {
```
Comment (ibzib) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/ClassLoaderFileSystem.java:107]: Can we add a couple trivial unit tests as sanity checks / documentation for `resolve` and `getCurrentDirectory`?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/ClassLoaderFileSystemTest.java (hunk 3 above)
Comment (robertwb) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/io/ClassLoaderFileSystemTest.java:74]: It's used on the Python side.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/ClassLoaderFileSystem.java (hunk 4 above)
Comment (robertwb) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/io/ClassLoaderFileSystem.java]: The documentation is in the super classes, but I added a test.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/ClassLoaderFileSystem.java (hunk 2 above)
Comment (robertwb) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/io/ClassLoaderFileSystem.java:39]: It's used on the Python side.

Comment (ibzib) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/io/ClassLoaderFileSystem.java:39]: That's what I get for only reading half the PR...

**Comment by robertwb** [comment]
Still trying to figure out why the test fails on jenkins but passes locally, but other than that it should be ready to be looked at again.

File: sdks/python/apache_beam/runners/portability/abstract_job_service.py [hunk 5]
```diff
@@ -313,39 +329,62 @@ def prepare(self):
...
     _LOGGER.info('Artifact server started on port %s', port)
+    _LOGGER.error('Artifact server started on port %s', port)
     return port
 
   def _stop_artifact_service(self):
     self._artifact_staging_server.stop(1)
-    self._artifact_staging_service.close()
-    self._artifact_manifest_location = (
-        self._artifact_staging_service.retrieval_token(self._job_id))
+    self._artifact_manifest_location = None
```
Comment (ibzib) [inline sdks/python/apache_beam/runners/portability/abstract_job_service.py:359]: Can't we remove this now?

File: sdks/python/apache_beam/runners/portability/abstract_job_service.py [hunk 6]
```diff
@@ -313,39 +329,62 @@ def prepare(self):
...
     port = self._artifact_staging_server.add_insecure_port(
         '[::]:%s' % requested_port)
-    beam_artifact_api_pb2_grpc.add_LegacyArtifactStagingServiceServicer_to_server(
+    beam_artifact_api_pb2_grpc.add_ArtifactStagingServiceServicer_to_server(
         self._artifact_staging_service, self._artifact_staging_server)
     self._artifact_staging_endpoint = endpoints_pb2.ApiServiceDescriptor(
         url='localhost:%d' % port)
     self._artifact_staging_server.start()
     _LOGGER.info('Artifact server started on port %s', port)
+    _LOGGER.error('Artifact server started on port %s', port)
```
Comment (ibzib) [inline sdks/python/apache_beam/runners/portability/abstract_job_service.py:354]: Why error?

File: sdks/python/apache_beam/runners/portability/abstract_job_service.py [hunk 7]
```diff
@@ -278,6 +279,23 @@ def to_runner_api(self):
...
 
+class JarArtifactManager(object):
+  def __init__(self, jar_path, root):
+    self._root = root
+    self._zipfile_handle = zipfile.ZipFile(jar_path, 'a')
+
+  def close(self):
+    self._zipfile_handle.close()
+
+  def file_writer(self, path):
```
Comment (ibzib) [inline sdks/python/apache_beam/runners/portability/abstract_job_service.py:290]: Can we get a type annotation and/or comment on the return value here?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/ClassLoaderFileSystem.java [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.io.fs.CreateOptions;
+import org.apache.beam.sdk.io.fs.MatchResult;
+import org.apache.beam.sdk.io.fs.ResolveOptions;
+import org.apache.beam.sdk.io.fs.ResourceId;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableList;
+
+/** A read-only {@link FileSystem} implementation looking up resources using a ClassLoader. */
+public class ClassLoaderFileSystem extends FileSystem<ClassLoaderFileSystem.ClassLoaderResourceId> {
```
Comment (ibzib) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/ClassLoaderFileSystem.java:40]: Should we change PortablePipelineJarUtils to use ClassLoaderFileSystem? (Maybe in a follow-up PR) https://github.com/apache/beam/blob/master/runners/java-job-service/src/main/java/org/apache/beam/runners/jobsubmission/PortablePipelineJarUtils.java

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/ClassLoaderFileSystem.java (hunk 4 above)
Comment (ibzib) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/io/ClassLoaderFileSystem.java]: Thanks!

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/ClassLoaderFileSystem.java (hunk 8 above)
Comment (robertwb) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/io/ClassLoaderFileSystem.java:40]: Yes, we'll want to do that too (and use the new artifact api).

File: sdks/python/apache_beam/runners/portability/abstract_job_service.py (hunk 7 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/portability/abstract_job_service.py:291]: Done.

File: sdks/python/apache_beam/runners/portability/abstract_job_service.py (hunk 6 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/portability/abstract_job_service.py]: Leftover debugging. Removed.

File: sdks/python/apache_beam/runners/portability/abstract_job_service.py (hunk 5 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/portability/abstract_job_service.py]: Yep. Done.

File: sdks/python/apache_beam/runners/portability/abstract_job_service.py [hunk 9]
```diff
@@ -288,6 +289,9 @@ def close(self):
     self._zipfile_handle.close()
 
   def file_writer(self, path):
+    # type: (str) -> Tuple[BinaryIO, str]
+    """Given a relative path, returns an open handle that can be written to
+    and an reference that can later be used to read this file."""
```
Comment (ibzib) [inline sdks/python/apache_beam/runners/portability/abstract_job_service.py:294]:
```suggestion
    and a reference that can later be used to read this file."""
```

**Pull request merged**